lib-content-scrapers/src/jvmMain/kotlin/com/ustadmobile/lib/contentscrapers/googledrive/GoogleDriveScraper.kt

```diff
@@ -0,0 +1,154 @@
+import com.ustadmobile.port.sharedse.contentformats.mimeTypeSupported
+import io.ktor.client.call.receive
+import io.ktor.client.request.get
+import io.ktor.client.request.parameter
+import io.ktor.client.statement.HttpStatement
+import kotlinx.coroutines.runBlocking
+import org.apache.commons.io.FileUtils
+import org.kodein.di.DI
+import org.kodein.di.instance
+import java.io.File
+import java.io.InputStream
+import java.nio.file.Files
+
+@ExperimentalStdlibApi
+class GoogleDriveScraper(contentEntryUid: Long, sqiUid: Int, parentContentEntryUid: Long, endpoint: Endpoint, di: DI) : Scraper(contentEntryUid, sqiUid, parentContentEntryUid, endpoint, di) {
+
+    private var tempDir: File? = null
+
+    val googleDriveFormat: DateFormat = DateFormat("yyyy-MM-dd'T'HH:mm:ss.SSSZ")
+
+    val googleApiKey: String by di.instance(tag = DiTag.TAG_GOOGLE_API)
+
+    override fun scrapeUrl(sourceUrl: String) {
+
+        var apiCall = sourceUrl
+        var fileId = sourceUrl.substringAfter("https://www.googleapis.com/drive/v3/files/")
+        if (sourceUrl.startsWith("https://drive.google.com/file/d/")) {
+            val fileIdLookUp = sourceUrl.substringAfter("https://drive.google.com/file/d/")
+            val char = fileIdLookUp.firstOrNull { it == '/' || it == '?' }
+            fileId = if (char == null) fileIdLookUp else fileIdLookUp.substringBefore(char)
+            apiCall = "https://www.googleapis.com/drive/v3/files/$fileId"
+        }
+
+        tempDir = Files.createTempDirectory(fileId).toFile()
+        runBlocking {
+
+            defaultHttpClient().get<HttpStatement>(apiCall) {
+                parameter("key", googleApiKey)
+                parameter("fields", "id,modifiedTime,name,mimeType,description,thumbnailLink")
+            }.execute() { fileResponse ->
+
+                val file = fileResponse.receive<GoogleFile>()
+
+                mimeTypeSupported.find { fileMimeType -> fileMimeType == file.mimeType }
+                        ?: return@execute
+
+                val recentContainer = containerDao.getMostRecentContainerForContentEntry(contentEntryUid)
+                val modifiedTime: Long = googleDriveFormat.parse(file.modifiedTime!!).local.unixMillisLong
+                val isUpdated = modifiedTime > recentContainer?.cntLastModified ?: 0
+                if (!isUpdated) {
+                    showContentEntry()
+                    setScrapeDone(true, 0)
+                    close()
+                    return@execute
+                }
+
+                defaultHttpClient().get<HttpStatement>(apiCall) {
+                    parameter("alt", "media")
+                    parameter("key", googleApiKey)
+                }.execute() {
+
+                    val contentFile = File(tempDir, file.name ?: file.id!!)
+                    val stream = it.receive<InputStream>()
+                    FileUtils.writeByteArrayToFile(contentFile, stream.readBytes())
```

Comment:
this would read the whole file into the memory. It would be better to stream it to disk. Even on the server that is best avoided (e.g. if handling larger video files etc).